lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -153,85 +184,35 @@ public void OnNext(IActiveContext activeContext)
         {
             Logger.Log(Level.Verbose, string.Format("Received Active Context {0}", activeContext.Id));
 
-            if (_groupCommDriver.IsMasterTaskContext(activeContext))
+            if (!_reachedUpdateTaskActiveContext)
             {
+                Logger.Log(Level.Info, "Submitting master task");
                 _reachedUpdateTaskActiveContext = true;
-                RequestMapEvaluators(_dataSet.Count);
-
-                var partialTaskConf =
-                    TangFactory.GetTang()
-                        .NewConfigurationBuilder(new[]
-                        {
-                            TaskConfiguration.ConfigurationModule
-                                .Set(TaskConfiguration.Identifier,
-                                    IMRUConstants.UpdateTaskName)
-                                .Set(TaskConfiguration.Task,
-                                    GenericType<UpdateTaskHost<TMapInput, TMapOutput, TResult>>.Class)
-                                .Build(),
-                            _configurationManager.UpdateFunctionConfiguration,
-                            _configurationManager.ResultHandlerConfiguration
-                        })
-                        .BindNamedParameter(typeof(InvokeGC), _invokeGC.ToString())
-                        .Build();
-
-                try
-                {
-                    TangFactory.GetTang()
-                        .NewInjector(partialTaskConf, _configurationManager.UpdateFunctionCodecsConfiguration)
-                        .GetInstance<IIMRUResultHandler<TResult>>();
-                }
-                catch (InjectionException)
-                {
-                    partialTaskConf = TangFactory.GetTang().NewConfigurationBuilder(partialTaskConf)
-                        .BindImplementation(GenericType<IIMRUResultHandler<TResult>>.Class,
-                            GenericType<DefaultResultHandler<TResult>>.Class)
-                        .Build();
-                    Logger.Log(Level.Warning,
-                        "User has not given any way to handle IMRU result, defaulting to ignoring it");
-                }
-
+                var groupCommConfig = GetGroupCommConfiguration();
                 _commGroup.AddTask(IMRUConstants.UpdateTaskName);
-                _groupCommTaskStarter.QueueTask(partialTaskConf, activeContext);
+                var finalTaskConfig = Configurations.Merge(groupCommConfig, UpdateTaskConfiguration());
+                _groupCommTaskStarter.QueueTask(finalTaskConfig, activeContext);
+                RequestMapEvaluators(_dataSet.Count);
             }
             else
             {
+                Logger.Log(Level.Info, "Submitting map task");
+                _serviceAndContextConfigurationProvider.ReachedActiveContext(activeContext.EvaluatorId);
+                var groupCommConfig = GetGroupCommConfiguration();
                 string taskId;
-
                 if (!_taskIdStack.TryPop(out taskId))
```

Comment:
Why use pre-created task id? As the task id is created in the driver constructor while partitionDescriptor is created, in fault tolerant case, we don't want to regenerate partitionDescriptor, but we do want to create new tasks, that makes decouple harder. 
Is there any relationship between taskid and partitionDescriptor ?